### PR TITLE
DDP-4925 replace app-engine-dispatch with load-balancer

### PR DESCRIPTION
**Pull request by yufengwng**
## Context

Here are the scripts that will help us setup the initial load-balancer to replace app engine `dispatch.yaml`. Start by reading the `init-lb.sh` script. The `add-study-to-lb.sh` script can be used down the line to more easily add LB configs for a new study. See this [wiki page](https://broadinstitute.atlassian.net/wiki/spaces/DDP/pages/2635497487/Dispatch+Load+Balancer) for more details.

CC @bskinner since I saw we have a LB for Salt so I could probably use some advice here.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

- [ ] :relaxed: All good, business as usual!
- [x] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

I have already setup the LB in `dev` environment so that it can be demo-ed there.

## Testing

Initial testing of the LB in `dev` did not show any issues. I was able to reach the various apps and services just fine.

## Release

TLDR: There will be downtime!

First, we need to run `init-lb.sh` to setup the load-balancer. Then, we can run `add-study-to-lb.sh` to add in all the various studies. We might need to do some custom/additional steps depending on the study.

The main source of downtime is switching DNS and provisioning SSL certs. From what I can tell in `dev`, updated DNS records seem to propagate pretty quickly, so downtime was mostly dominated by SSL cert provisioning which took anywhere from 8 - 12 mins. So I think a rough estimate of downtime is about 15 mins.

We should be able to apply this without having to create a new code release.

### CODE REVIEW

File: scripts/load-balancer/add-study-to-lb.sh [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+echo "using lb name: $LB_NAME"
+echo "using gae service: $GAE_SERVICE"
+echo "using study domain: $STUDY_DOMAIN"
+echo "using study identifier: $STUDY"
+echo ""
+
+echo_run() {
+  echo "+ $@"
+  "$@"
+  echo ""
```
Comment (bskinner) [inline scripts/load-balancer/add-study-to-lb.sh:34]: Would using `set -v` or `set -x` before line 33 and disabling the option (`set +<opt>`) be an option here instead of `echo`, or does the output need to be on `stdout`?

File: scripts/load-balancer/init-lb.sh [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+echo "=> setting up second LB for http-to-https redirects..."
+
+# Note: we use another LB to handle http-to-https redirects. The second LB
+# mostly just contains frontends with port 80 and same IPs, and no backends.
+# There doesn't seem to be a way to create this using cli interface, so we have
+# to import a YAML file.
+#
+# We're only going to use this http LB for redirecting the study site and not
+# our DSS and DSM backends. HTTP requests might contain credentials, so we want
+# to reject those requests.
```
Comment (bskinner) [inline scripts/load-balancer/init-lb.sh:117]: That's a nice comment 👍

**Review by bskinner — APPROVED** [review]
Had a question regarding the `echo_run` function, but looks good otherwise!

**Comment by bskinner** [comment]
Did some digging into using Cloud Armor for blocking the `http` routes explicitly instead of implicitly `404`ing them. Your approach here works great, and achieves the same end. I think we may get some benefit out of explicitly blocking those routes when `http` is used, but that's something to look at later.

File: scripts/load-balancer/add-study-to-lb.sh (hunk 1 above)
Comment (yufengwng) [reply scripts/load-balancer/add-study-to-lb.sh:34]: Doesn't have to echo to stdout. I can use `set -x` here.

Comment (yufengwng) [reply scripts/load-balancer/add-study-to-lb.sh:34]: Hmm nevermind, `set -x` is not what I want here. I think I'll stick with `echo`. Using `set -x` will echo every command, so it means it will also echo the `set +x` after executing the command, resulting in something like:
```
+ gcloud compute ...
+ set +x

+ gcloud compute ...
+ set +x
```

Comment (yufengwng) [reply scripts/load-balancer/add-study-to-lb.sh:34]: Also, I believe [the docs](https://tldp.org/LDP/abs/html/options.html) is saying both `-v` and `-x` prints to stdout. So I guess my approach with `echo` is pretty much the same in that regard.
